File: 1283. Find the Smallest Divisor Given a Threshold.js
// INPUT: [ 1, 2, 5, 9 ], 6
// OUTPUT: 5

const isBelowThreshould = (nums, threshold, divisor) => {
  let sum = 0;

  for (let i = 0; i < nums.length; i++) {
    sum += Math.ceil(nums[i] / divisor);
  }

  return sum <= threshold;
};

const smallestDivisor = (nums, threshold) => {
  let left = 1;
  let right = Math.max(...nums);

  while (left < right) {
    const mid = Math.floor((left + right) / 2);

    if (isBelowThreshould(nums, threshold, mid)) {
      right = mid;
    } else {
      left = mid + 1;
    }
  }

  return left;
};

console.log(smallestDivisor([1, 2, 5, 9], 6));

File: python/1283. Find the Smallest Divisor Given a Threshold.py
from typing import List
import math


class Solution:
    def smallestDivisor(self, nums: List[int], threshold: int) -> int:
        left = 1
        right = max(nums)

        while left < right:
            mid = (left + right) // 2

            curr_sum = 0

            for num in nums:
                curr_sum += math.ceil(num / mid)

            if curr_sum <= threshold:
                right = mid
            else:
                left = mid + 1

        return left
